### PR TITLE
Task 4/test service layer

### PR DESCRIPTION
**Pull request by Orkhan93**

### CODE REVIEW

File: src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/ErrorHandler.java [hunk 1]
```diff
@@ -20,6 +22,7 @@ public ErrorResponse handleCustomException(ServiceException ex) {
         ErrorResponse errorResponse = new ErrorResponse();
         errorResponse.setCode(ex.getCode());
         errorResponse.setMessage(ex.getMessage());
+        log.info("Service Exception : {}", ex.getMessage());
```
Comment (Allahverdiyev-Rashad) [inline src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/ErrorHandler.java:25]: log.error

File: src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/ErrorHandler.java [hunk 2]
```diff
@@ -33,6 +36,7 @@ public ErrorResponse handleValidationException(MethodArgumentTypeMismatchExcepti
         ErrorResponse errorResponse = new ErrorResponse();
         errorResponse.setCode(ErrorCode.VALIDATION_ERROR.name());
         errorResponse.setMessage(parameterName + ErrorMessage.VALIDATION_ERROR);
+        log.info("Validation server error : {}", ex.getMessage());
```
Comment (Allahverdiyev-Rashad) [inline src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/ErrorHandler.java:39]: log.error

File: src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/ErrorHandler.java [hunk 3]
```diff
@@ -43,18 +47,20 @@ public ErrorResponse handleEmptyResultDataAccess(EmptyResultDataAccessException
         ErrorResponse errorResponse = new ErrorResponse();
         errorResponse.setCode(ErrorCode.VALIDATION_ERROR.name());
         errorResponse.setMessage(ErrorMessage.STUDENT_NOT_FOUND);
+        log.info("Validation server error : {}", ex.getMessage());
```
Comment (Allahverdiyev-Rashad) [inline src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/ErrorHandler.java:50]: log.error

File: src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/ErrorHandler.java [hunk 4]
```diff
@@ -43,18 +47,20 @@ public ErrorResponse handleEmptyResultDataAccess(EmptyResultDataAccessException
...
-    public ErrorResponse handleArgumentNotValidException(MethodArgumentNotValidException ex){
+    public ErrorResponse handleArgumentNotValidException(MethodArgumentNotValidException ex) {
 
-       String fieldName = ex.getBindingResult().getFieldError().getField();
+        String fieldName = ex.getBindingResult().getFieldError().getField();
 
         ErrorResponse errorResponse = new ErrorResponse();
         errorResponse.setCode(ErrorCode.VALIDATION_ERROR.name());
         errorResponse.setMessage(fieldName + ErrorMessage.VALIDATION_ERROR);
+        log.info("Validation error : {}", ex.getMessage());
```
Comment (Allahverdiyev-Rashad) [inline src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/ErrorHandler.java:63]: log.error

File: src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/ErrorHandler.java [hunk 5]
```diff
@@ -64,6 +70,7 @@ public ErrorResponse handleUnknownException(Exception e) {
         ErrorResponse errorResponse = new ErrorResponse();
         errorResponse.setCode(ErrorCode.INTERNAL_SERVER_ERROR.name());
         errorResponse.setMessage(ErrorMessage.INTERNAL_SERVER_ERROR);
+        log.info("Internal server error : {}", e.getMessage());
```
Comment (Allahverdiyev-Rashad) [inline src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/ErrorHandler.java:73]: log.error

File: src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/constraint/validation/Name.java [hunk 6]
```diff
@@ -18,6 +18,8 @@
 @NotBlank(message = "Must not be blank")
 public @interface Name {
 
+    String message() default "{}";
```
Comment (Allahverdiyev-Rashad) [inline src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/constraint/validation/Name.java:21]: no need {} only ""

File: src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/constraint/validation/Surname.java [hunk 7]
```diff
@@ -18,6 +18,8 @@
 @NotBlank(message = "Must not be blank")
 public @interface Surname {
 
+    String message() default "{}";
```
Comment (Allahverdiyev-Rashad) [inline src/main/java/az/spring/studentmanagementservice/studentmanagementservice/error/constraint/validation/Surname.java:21]: no need {} only ""

File: src/main/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentService.java [hunk 8]
```diff
@@ -39,22 +39,24 @@ public StudentResponse getStudentById(Long studentId) {
         return studentResponse;
     }
 
-    public void createStudent(StudentRequest studentRequest) {
+    public StudentRequest createStudent(StudentRequest studentRequest) {
```
Comment (Allahverdiyev-Rashad) [inline src/main/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentService.java:42]: always return StudentResponse. use two different dto StudentRequest and StudentResponse

File: src/main/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentService.java [hunk 9]
```diff
@@ -39,22 +39,24 @@ public StudentResponse getStudentById(Long studentId) {
...
         log.info("studentRequest : {}", studentRequest);
         Student student = new Student();
         BeanUtils.copyProperties(studentRequest, student);
         studentRepository.save(student);
         log.info("student : {}", student);
+        return studentRequest;
     }
 
-    public void updateStudent(Long studentId, StudentRequest studentRequest) {
+    public StudentRequest updateStudent(Long studentId, StudentRequest studentRequest) {
```
Comment (Allahverdiyev-Rashad) [inline src/main/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentService.java:51]: same here

File: src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class StudentServiceTest {
```
Comment (Allahverdiyev-Rashad) [inline src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java:30]: after correct return values update test methods as well

File: src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+    @Test
+    public void testDeleteStudent() {
+        studentRepository.deleteById(2L);
+        verify(studentRepository, times(1)).deleteById(anyLong());
+        verifyNoMoreInteractions(studentRepository);
+    }
+
+    @Test
+    public void testException_whenStudentIdDoesNotExist_shouldThrowStudentNotFoundException() {
```
Comment (Allahverdiyev-Rashad) [inline src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java:146]: write methodname_when..._shouldThrow...

File: src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        try {
```
Comment (Allahverdiyev-Rashad) [inline src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java:147]: not like this. mock repo and when(repo.method).thenThrow()

File: src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java (hunk 10 above)
Comment (Orkhan93) [reply src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java:30]: Could you please provide more detailed information

File: src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+        StudentListResponse resultList = studentService.getAllStudents();
+        assertThat(resultList).isNotNull();
+        assertThat(resultList.getStudents());
+
+        verify(studentRepository).findAll();
+        verifyNoMoreInteractions(studentRepository);
+    }
+
+    //not found
```
Comment (Allahverdiyev-Rashad) [inline src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java:115]: remove redundant comment

File: src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+        when(studentRepository.findById(student.getId())).thenReturn(Optional.of(student));
+
+        StudentResponse result = studentService.updateStudent(student.getId(), studentRequest);
+        assertThat(result).isNotNull();
+        assertEquals(studentRequest.getId(), student.getId());
+    }
+
+    @Test
+    public void testDeleteStudent() {
+        studentRepository.deleteById(2L);
```
Comment (Allahverdiyev-Rashad) [inline src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java:139]: where is service call?

File: src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+        verifyNoMoreInteractions(studentRepository);
+    }
+
+    @Test
+    public void testHandleCustomException_whenStudentIdDoesNotExist_shouldThrowStudentNotFoundException() {
+        when(studentRepository.findById(anyLong())).thenReturn(Optional.empty());
+
+        assertThrows(ServiceException.class, () -> studentService.getStudentById(anyLong()));
+
+        verify(studentRepository).findById(anyLong());
```
Comment (Allahverdiyev-Rashad) [inline src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java:150]: verify no more interaction

File: src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+
+    @Test
+    public void testDeleteStudent() {
+        studentRepository.deleteById(2L);
+        verify(studentRepository, times(1)).deleteById(anyLong());
+        verifyNoMoreInteractions(studentRepository);
+    }
+
+    @Test
+    public void testHandleCustomException_whenStudentIdDoesNotExist_shouldThrowStudentNotFoundException() {
```
Comment (Allahverdiyev-Rashad) [inline src/test/java/az/spring/studentmanagementservice/studentmanagementservice/service/StudentServiceTest.java:145]: change name to tested method name

**Review by Allahverdiyev-Rashad — COMMENTED** [review]
write all tests with 100 percent covarage. right click test class and run with covarage

**Review by Allahverdiyev-Rashad — APPROVED** [review]

**Pull request merged**